code-dot-org/code-dot-org
# [Google Blockly] CT-133: Handle a project that can switch on and off the modal function editor

The following PR: 
- Modifies getCode to merge the serializations for both the main and hidden workspaces and return a combined serialization 
- Adds code to reset the block configuration that gets added in the editor workspace (a fixed x and y position and immovability) before we save the hidden workspace serialization 
- Modifies `moveHiddenProcedures` to move blocks from the base source to a blank/newly populated hiddenDefinitionSource based on the procedureTypesToHide each time loadBlocksToWorkspace is called
- Removes all references to hiddenDefinitions in `StudioApp.js` and `project.js`

Unrelated/tangentially related changes:
- Adds a new positioning algorithm that will handle a list of blocks with a mix of set and unset positions

TODOs:
- [x] Ask about code saving (Is it possible this PR is having an impact on frequency?)
- [x] Check on whether it's okay to not be able to delete the definitions on the main workspace
- [x] Test with a level with the function editor turned on and then turned off
- [x] Should a new algorithm for positioning be introduced in a follow-up PR?

## Links

Jira ticket: https://codedotorg.atlassian.net/browse/CT-133
Original Slack thread: https://codedotorg.slack.com/archives/C05DK21DAHK/p1697497810140919

## Testing story

1. Added a functions category to the toolbox of level courseE_aboutme_template_2023
2. Edited level courseE_aboutme_3_2023 to disable the modal function editor (other levels in this progression have it disabled)
3. Used the modal function editor on http://localhost-studio.code.org:3000/s/coursee-2023/lessons/5/levels/2?blocklyVersion=Google to define a new function (in screenshots below, `becomeHulk`)
4. Ensure code is saved and move on to http://localhost-studio.code.org:3000/s/coursee-2023/lessons/5/levels/3?blocklyVersion=Google
5. Ensure that the function renders/is auto-positioned correctly on the main workspace when the modal editor is disabled and that the block can be moved
6. Check that on http://localhost-studio.code.org:3000/s/coursee-2023/lessons/5/levels/4?blocklyVersion=Google, when the modal function editor is reenabled, that the function definition is hidden and can be edited with the modal function editor

Note: I have not yet tested a level with the function editor turned on and then turned off.

<img width="1019" alt="Screenshot 2023-11-01 at 9 17 29 PM" src="https://github.com/code-dot-org/code-dot-org/assets/26844240/ce826ba6-e3a6-4c00-8965-47829119dec1">

<img width="1020" alt="Screenshot 2023-11-01 at 9 18 06 PM" src="https://github.com/code-dot-org/code-dot-org/assets/26844240/bf9d7a57-38d5-4e71-ab3e-97cbc9f3e6a2">

**Block positioning screenshots**
On a LTR workspace
<img width="401" alt="Screenshot 2023-11-07 at 10 42 13 PM" src="https://github.com/code-dot-org/code-dot-org/assets/26844240/0671b742-d40a-4543-9e74-8fc7f3e7f4c2">

On a RTL workspace
<img width="1124" alt="Screenshot 2023-11-07 at 10 44 29 PM" src="https://github.com/code-dot-org/code-dot-org/assets/26844240/da5e4926-6f75-4990-99b1-68e45c73a4ca">

## Deployment strategy

## Follow-up work



## Privacy



## Security



## Caching

## PR Checklist:
